src/Backend/Haskell.idr

```diff
@@ -25,12 +26,69 @@ data HsType : Type where -- TODO could be interesting to index this by e.g. used
   ||| The tuple type, containing two or more further types.
   HsTuple :         Vect (2 + n) HsType -> HsType
 
+  ||| The sum type, containing two further types.
+  HsSum   :         HsType -> HsType    -> HsType
+
   ||| A type variable.
   HsVar   :         Name                -> HsType
 
   ||| A named type with zero or more other types as parameters.
   HsParam : Name -> Vect n HsType       -> HsType
 
+  ||| A function type (only used for top-level type of termdef
+  ||| decoders and encoders)
+  HsArrow :         HsType -> HsType    -> HsType
+
+private
+hsParam : Decl -> HsType
+hsParam (MkDecl n ps) = HsParam n (map HsVar ps)
+
+||| The syntactic structure of (a subset of) Haskell terms.
+data HsTerm : Type where
+  ||| The unit term `()`.
+  HsUnitTT  : HsTerm
```

Comment:
It would probably be nice to align things here as in the `HsType` definition

Reply:
Done! (and soon pushed)